### PR TITLE
Rewrite the only place dependent on `typhoeus`

### PR DESCRIPTION
**Pull request by jasl**
`typhoeus` is the only blocking that prevent CocoaPods running on M1 natively

`cdn_url?` is the only place that using `typhoeus` so I just rewrite it with Ruby's `OpenURI`

WebMock dependency was updated in the `Gemfile.lock` to support testing natively bundled `OpenURI` on newer Ruby versions.

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* replace typhoeus with open-uri  
  [jasl](https://github.com/jasl)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by jasl** [comment]
CI green now

**Comment by jasl** [comment]
Add a comment about why `URI.send :open`

File: CHANGELOG.md
Comment (igor-makarov) [inline CHANGELOG.md:11]: nit: please add 2 spaces at the end of line

File: lib/cocoapods/sources_manager.rb
Comment (igor-makarov) [inline lib/cocoapods/sources_manager.rb:80]: Perhaps name `netrc_info`?

File: lib/cocoapods/sources_manager.rb
Comment (igor-makarov) [inline lib/cocoapods/sources_manager.rb:82]: nit: `credentials`

File: lib/cocoapods/sources_manager.rb
Comment (igor-makarov) [inline lib/cocoapods/sources_manager.rb:83]: nit: `uri_options`

**Comment by igor-makarov** [comment]
@jasl I've replaced the call to private method to `OpenURI::open_uri` - this is both portable and readable.

I've left some comments as part of the review. 

Also, I believe that if you remove the `Gemfile` and `gemspec` changes, we could merge this sooner.

Removing Typhoeus will require Core update first anyway, as it's a transitive dependency.

**Comment by jasl** [comment]
> @jasl I've replaced the call to private method to `OpenURI::open_uri` - this is both portable and readable.
> 
> I've left some comments as part of the review.
> 
> Also, I believe that if you remove the `Gemfile` and `gemspec` changes, we could merge this sooner.
> 
> Removing Typhoeus will require Core update first anyway, as it's a transitive dependency.

You mean I need to keeping `typhoeus` as CocoaPods dependency? but if `cocoapods-core` dependent `typhoeus`, `cocoapods` should also dependent it indirectly?

in Gemfile.lock you can see

```
GIT
  remote: https://github.com/CocoaPods/Core.git
  revision: f4d4f16f0d92f422f15b1f27e2073276098f8f73
  branch: master
  specs:
    cocoapods-core (1.10.1)
      activesupport (> 5.0, < 6)
      addressable (~> 2.6)
      algoliasearch (~> 1.0)
      concurrent-ruby (~> 1.1)
      fuzzy_match (~> 2.0.4)
      nap (~> 1.0)
      netrc (~> 0.11)
      public_suffix
      typhoeus (~> 1.0)
```

so `typhoeus` still on the list

**Comment by jasl** [comment]
@igor-makarov 
applied your suggests, thanks!

**Comment by igor-makarov** [comment]
@jasl Yes - what I mean is that there's no benefit in removing the dependency now, as it will still get installed through Core.

File: CHANGELOG.md
Comment (igor-makarov) [inline CHANGELOG.md:14]: @jasl I'm sorry I wasn't clear - the two spaces are supposed to be at the end of line 11, not 14 😂

**Comment by jasl** [comment]
> @jasl Yes - what I mean is that there's no benefit in removing the dependency now, as it will still get installed through Core.

I still don't understand why should keeping it in `CocoaPods`... but I just do as your sugguest.

I squash commits, added your credit in Changelog, thanks you for reviewing!

**Comment by igor-makarov** [comment]
Looks good - however there's still a `Gemfile.lock` change.

**Comment by jasl** [comment]
Because we don't remove dependency of `typhoeus`, I change the title and changelog more accurate

**Comment by jasl** [comment]
> @jasl Looks good - however there's still a `Gemfile.lock` change.

I have to upgrade `webmock` because it doesn't support newer Rubies bundled `open-uri`

**Comment by igor-makarov** [comment]
@jasl got it now - I updated the PR description to mention this.

**Review by igor-makarov — APPROVED** [review]

**Comment by jasl** [comment]
> @jasl got it now - I updated the PR description to mention this.

Thank you!

File: lib/cocoapods/sources_manager.rb
Comment (segiddins) [inline lib/cocoapods/sources_manager.rb:77]: out of curiosity, why disable redirects?

File: lib/cocoapods/sources_manager.rb
Comment (segiddins) [inline lib/cocoapods/sources_manager.rb:89]: can rescue the three error classes in the same rescue statement

**Review by segiddins — APPROVED** [review]

File: lib/cocoapods/sources_manager.rb
Comment (jasl) [reply lib/cocoapods/sources_manager.rb:77]: the previous implemention

```ruby
response.code == 200 && begin
            response_hash = YAML.load(response.body) # rubocop:disable Security/YAMLLoad
            response_hash.is_a?(Hash) && !Source::Metadata.new(response_hash).latest_cocoapods_version.nil?
          end
```

and there's a test https://github.com/CocoaPods/CocoaPods/blob/master/spec/unit/sources_manager_spec.rb#L181-L185

I just keeping the same behavior

File: lib/cocoapods/sources_manager.rb
Comment (jasl) [reply lib/cocoapods/sources_manager.rb:89]: will do

File: lib/cocoapods/sources_manager.rb
Comment (amorde) [reply lib/cocoapods/sources_manager.rb:77]: the linked test:
```
        it 'redirect' do
          WebMock.stub_request(:get, 'http://some_host.com/something/CocoaPods-version.yml').
            to_return(:status => 301, :body => '', :headers => { 'Location' => ['http://some_host.com'] })
          @sources_manager.cdn_url?('http://some_host.com/something').should == false
        end
```

Isn't testing that redirects are _not_ followed. It's testing that if `some_host.com/something/CocoaPods-version.yml` redirects to `some_host.com/`, then the source in question isn't a CDN. I might be misunderstanding this, but I don't think we want to disable redirects

Comment (jasl) [reply lib/cocoapods/sources_manager.rb:77]: > It's testing that if some_host.com/something/CocoaPods-version.yml redirects to some_host.com/, then the source in question isn't a CDN.

In my understanding, `cdn_url?` attempting to get `"#{given_url}/CocoaPods-version.yml"` and trying to parse it, then it returns true.

base on your explain, you mean `cdn_url?` getting `http://some_host.com/something/CocoaPods-version.yml` through redirection returns `http://some_host.com/something`, in the test, it's not mocked, so maybe returns `SocketError`, so it's not a `cdn_url` then returns `false`

is this what your means?

Comment (jasl) [reply lib/cocoapods/sources_manager.rb:77]: I'm not consider the logic before... my mistake, your explain sounds reasonable to me

Comment (jasl) [reply lib/cocoapods/sources_manager.rb]: @amorde @segiddins 
Sorry, that's my mistake, I misunderstanding the meaning of that test,

I've corrected it

Comment (amorde) [reply lib/cocoapods/sources_manager.rb]: Apologies on the unclear wording. I think that test is actually relying on a redirect, and expects redirects to be supported and not disabled. I will defer to @igor-makarov to confirm the intent behind the test

Comment (jasl) [reply lib/cocoapods/sources_manager.rb]: I enable redirects now, looks CI green.

**Comment by jasl** [comment]
@igor-makarov

**Comment by jasl** [comment]
@igor-makarov 
I pushed some changes:

- Re-enable redirects, see conversation above, that's my mistake
- Revert your change `URI.join(url, "CocoaPods-version.yml")` becasue
  ```
  URI.join("http://some_host.com/something", "CocoaPods-version.yml")
   => #<URI::HTTP http://some_host.com/CocoaPods-version.yml>
  ```
- Add an unit test about `netrc`

File: spec/unit/sources_manager_spec.rb
Comment (jasl) [inline spec/unit/sources_manager_spec.rb:197]: I think this is needed, or `open-uri` will trying to request to the Internet,
can proof this by `WebMock.disable_net_connect!`

File: lib/cocoapods/sources_manager.rb
Comment (igor-makarov) [reply lib/cocoapods/sources_manager.rb]: Yes, this is the test case. Perhaps it could be clearer worded. 🤔

File: spec/unit/sources_manager_spec.rb
Comment (igor-makarov) [reply spec/unit/sources_manager_spec.rb:197]: Agreed.

**Comment by igor-makarov** [comment]
Regarding `URI#join`, you are correct about this and I've found the [path vs. endpoint](https://apidock.com/ruby/URI/join/class#1245-Be-careful-with-path-vs-endpoint) warning on APIdock comments.

Thus, I believe that if you use 
```
URI.join(url + '/', "CocoaPods-version.yml")
```

- then we'll get the behavior we want with the slashes on or off. Might wanna add a test for it though.

**Comment by jasl** [comment]
I just add more tests for tailing `/` and `handle redirects`

I have to leave about 2 hours, when I back to office, I'll try `URI.join(url + '/', "CocoaPods-version.yml")`

File: CHANGELOG.md
Comment (dnkoutso) [inline CHANGELOG.md:14]: nit missing empty line above this.

**Comment by jasl** [comment]
> Regarding `URI#join`, you are correct about this and I've found the [path vs. endpoint](https://apidock.com/ruby/URI/join/class#1245-Be-careful-with-path-vs-endpoint) warning on APIdock comments.
> 
> Thus, I believe that if you use
> 
> ```
> URI.join(url + '/', "CocoaPods-version.yml")
> ```
> 
> then we'll get the behavior we want with the slashes on or off. Might wanna add a test for it though.
> 
> The reason I'm hung up on this is because people tend to sometimes add `source "https://cdn.cocoapods.org"` and sometimes `source "https://cdn.cocoapods.org/"` and we don't want either one to fail.

Applied this suggest, I've add tests for with or without tailing `/`

**Comment by jasl** [comment]
```
LoadError: dlopen(/Users/runner/work/CocoaPods/CocoaPods/vendor/bundle/gems/bigdecimal-1.3.5/lib/bigdecimal.bundle, 9): Library not loaded: /usr/local/opt/ruby/lib/libruby.2.7.dylib
  Referenced from: /Users/runner/work/CocoaPods/CocoaPods/vendor/bundle/gems/bigdecimal-1.3.5/lib/bigdecimal.bundle
```

Not the code problem, don't know how to restart it, but I believe it should no problem on Ruby 2.7

**Comment by igor-makarov** [comment]
@jasl @dnkoutso I think I know what's up. I've decided to add macOS 11 to the system test matrix #10350 and while adding it, I realized that the bundler cache keys aren't differentiated by OS version 🤦🏻‍♂️ but not before probably clobbering the bundler cache with newer native gems.

I think this PR will need to be rebased on main after #10350 is merged. Also, this means that we'll need to upstream #10350 into stable branch in case there would be PRs there.

**Comment by jasl** [comment]
> @jasl @dnkoutso I think I know what's up. I've decided to add macOS 11 to the system test matrix #10350 and while adding it, I realized that the bundler cache keys aren't differentiated by OS version 🤦🏻‍♂️ but not before probably clobbering the bundler cache with newer native gems.
> 
> I think this PR will need to be rebased on main after #10350 is merged. Also, this means that we'll need to upstream #10350 into stable branch in case there would be PRs there.



> @jasl @dnkoutso I think I know what's up. I've decided to add macOS 11 to the system test matrix #10350 and while adding it, I realized that the bundler cache keys aren't differentiated by OS version 🤦🏻‍♂️ but not before probably clobbering the bundler cache with newer native gems.
> 
> I think this PR will need to be rebased on main after #10350 is merged. Also, this means that we'll need to upstream #10350 into stable branch in case there would be PRs there.

Could you help to take a look https://github.com/CocoaPods/CocoaPods/pull/10347 ?

I'll waiting #10350 merge, now I need to find a safe way for URI appending, `URI.join(url + '/', "CocoaPods-version.yml")` this I believe it should work in real-word but not work for `webmock`

**Comment by igor-makarov** [comment]
I've looked at  #10347 and it looks good to me, but the custom JSON gem happened before my time, so I am waiting for more relevant people to take a look at it.

Regarding  #10350, I'm no longer certain that it's the cause for the failures. Investigating.

**Comment by igor-makarov** [comment]
We're gonna put adding Big Sur on hold for now, the PR that's blocking this PR is #10351 - which (hopefully) fixes the CI problems.

**Comment by jasl** [comment]
Rebased, waiting CI

**Comment by igor-makarov** [comment]
@igorudovika can you take a look at the changes to see if the NetRC stuff looks good?

File: lib/cocoapods/sources_manager.rb
Comment (jasl) [inline lib/cocoapods/sources_manager.rb:83]: `URI.join(url.chomp('/') + '/', 'CocoaPods-version.yml')` why not just `url.chomp('/') + '/', 'CocoaPods-version.yml'`?

Comment (igor-makarov) [reply lib/cocoapods/sources_manager.rb:83]: @jasl you're right, using `URI.join` seems redundant and cumbersome.

Comment (jasl) [reply lib/cocoapods/sources_manager.rb]: changed

File: spec/unit/sources_manager_spec.rb
Comment (igor-makarov) [inline spec/unit/sources_manager_spec.rb:31]: @jasl probably should be as above

**Comment by khomchyk** [comment]
Hello!
Can we, please, speed up the review of this PR? 
Seems that [#10351](https://github.com/CocoaPods/CocoaPods/pull/10351) has been already merged.

**Comment by igor-makarov** [comment]
@khomchyk CocoaPods is maintained by volunteers in their free time. There's no concept of "speeding up" when people volunteer.

That said, unblocking native M1 support might take a while as the current PR only covers one of the gems. There's a much bigger PR [here](https://github.com/CocoaPods/Core/pull/669) that will also need to be merged and it may take significant time. 

If you wanna use CocoaPods on M1 right now, please refer to #9907 for workarounds.

**Comment by igorudovika** [comment]
> @igorudovika can you take a look at the changes to see if the NetRC stuff looks good?

Sorry for late response. Sure, will look nearest time.

**Comment by igor-makarov** [comment]
@dnkoutso this can merged - I believe you've gone over it already?

**Comment by igor-makarov** [comment]
@jasl could you rebase please?

**Comment by jasl** [comment]
> @jasl could you rebase please?

Done.

**Review by dnkoutso — APPROVED** [review]

**Comment by im-ttmskk** [comment]
I am very much looking forward to the release of this contribution.

**Pull request merged**
